Sources/InternxtSwiftCore/Services/Network/Download.swift

```diff
@@ -0,0 +1,104 @@
+//
+//  File.swift
+//  
+//
+//  Created by Robert Garcia on 23/8/23.
+//
+
+import Foundation
+
+
+enum DownloadError: Error {
+    case DownloadNotSuccessful
+    case MissingDownloadURL
+    case MultipartDownloadNotSupported
+    case FailedToCopyDownloadedURL
+    case InvalidBucketId
+}
+
+public struct DownloadResult {
+    public var url: URL
+    public var expectedContentHash: String
+    public var index: String
+    init(url: URL, expectedContentHash: String, index: String) {
+        self.url = url
+        self.expectedContentHash = expectedContentHash
+        self.index = index
+    }
+}
+
+
+
+@available(macOS 10.15, *)
+public class Download: NSObject {
+    private let networkAPI: NetworkAPI
+    
+    private var observation: NSKeyValueObservation?
+
+    private var progressHandlersByTaskID = [Int : ProgressHandler]()
+    private var urlSession: URLSession
+    init(networkAPI: NetworkAPI, urlSession: URLSession?) {
+        self.networkAPI = networkAPI
+        if let urlSession = urlSession {
+            self.urlSession = urlSession
+        } else {
+            self.urlSession = URLSession.shared
+        }
+        
+        super.init()
+    }
+    
+    deinit {
+        observation?.invalidate()
+    }
+    
+    
+    func start(bucketId: String, fileId: String, destination: URL,  progressHandler: ProgressHandler? = nil,  debug: Bool = false) async throws ->  DownloadResult {
+        if bucketId.isValidHex == false {
+            throw DownloadError.InvalidBucketId
+        }
+        let info = try await networkAPI.getFileInfo(bucketId: bucketId, fileId: fileId)
+        
+        if info.shards.count > 1 {
```

Comment:
This is going to fail for uploads done before July - August 21'. Joining shards is not multipart (at least not typical multipart, as it does not use the common headers for that). If you receive multiple shards, just download and decrypt them (using the same decryptor) in order. In pseudocode: 

```
const decryptor = new DecryptStream();
const out = new FileSystemWriter();
for (const shard of shards) {
    const stream = getDownloadStream(shard)
   
    while (stream.isNotFinished) {
        decryptor.write(stream.read(<some_chunk_size>));
        out.write(decryptor.read(<some_chunk_size>));
     }
}
```

Reply:
Btw, I do not know if this scope should be tackled now, just saying.

Reply:
I will tackle legacy downloads and multipart in a separated task so it doesn't block the main development